### ISSUE TITLE
Not able to access API when sending Authorization Header with Access Token

### ISSUE THREAD

**Issue by mdmoura** [openiddict/openiddict-core#1327]
With Asp.Net Core 5, Angular and OpenIdDict I created 3 applications:

    Auth with OpenIdDict running on https://localhost:5000 [Asp.Net Core with OpenIdDict]
    API running on https://localhost:5001 [Asp.Net Core with OpenIdDict]
    SPA running on https://localhost:5002 [Asp.Net Core with Angular]

I am able to login and logout. The only problem is when sending the Authoization Header on an API endpoint call.

I am able to call API's endpoints but when I send the Authorization header with the Bearer Access Token I get the error:
 
    Access to XMLHttpRequest at 'https://localhost:5001/v1.0/posts' from origin 'https://localhost:5002' has been blocked by CORS policy: 
    
    No 'Access-Control-Allow-Origin' header is present on the requested resource.

    POST https://localhost:5001/v1.0/posts

I have been changing the configuration (adding `AllowCredentials()`, etc) and its order but I always get the same result ...

Is the order of OpenIdDict configuration relevant? Is this a bug? Am I missing something?

**My Angular's request:**

    httpClient.post(`https://localhost:5001/v1.0/posts`, 
     { title: "My post", body: "Some text" }, 
     { headers: { 
         'Content-Type': 'application/json', 
         'Authorization': `Bearer ${user.access_token}` 
       } 
     }).subscribe();

The **API Startup's** `ConfigureServices` and `Configure` methods are:

    public void ConfigureServices(IServiceCollection services) {

      services
        .AddControllers()
        .SetCompatibilityVersion(CompatibilityVersion.Latest)
        .AddJsonOptions()
        .AddFluentValidation();

      services
        .AddApiVersioning(x => {
          x.ApiVersionSelector = new CurrentImplementationApiVersionSelector(x);
          x.AssumeDefaultVersionWhenUnspecified = true;
          x.DefaultApiVersion = new ApiVersion(1, 0);
          x.ReportApiVersions = true;
          x.RouteConstraintName = "version";
        });

      services.AddDbContext<Context>(x =>
        x.UseSqlServer(Configuration.Get<Options>().Database.Connection, y => {
          y.MigrationsHistoryTable("__Migrations");
          y.UseNetTopologySuite();
        }).EnableSensitiveDataLogging(Environment.IsDevelopment())
          .UseQueryTrackingBehavior(QueryTrackingBehavior.TrackAll));

      services.AddRouting(x => { x.AppendTrailingSlash = false; x.LowercaseUrls = true; });

      services.AddCors(x => {
        x.AddPolicy("Cors", y => 
          y.WithOrigins("https://localhost:5000", "https://localhost:5002").AllowAnyMethod().AllowAnyHeader().AllowCredentials());
        });

      services.AddAuthentication(OpenIddictValidationAspNetCoreDefaults.AuthenticationScheme);

      services
        .AddAuthorization(x => {
          x.AddPolicy(Policy.Admin, y => y.RequireClaim("Admin"));
        });

      services
        .AddIdentityCore<User>(x => x.AddDefaultOptions())
        .AddEntityFrameworkStores<Context>()
        .AddDefaultTokenProviders();

      services.AddOpenIddict()
        .AddValidation(x => {
          x.SetIssuer("https://localhost:5000");
          x.AddAudiences("api");
          x.UseIntrospection();
          x.SetClientId("api").SetClientSecret("test");
          x.UseSystemNetHttp();
          x.UseAspNetCore();
        });

      services.AddHsts();
      services.AddHttpsRedirection();

      services.AddMediatR(typeof(Startup));

      services.Configure<ApiBehaviorOptions>(x => {
        x.SuppressModelStateInvalidFilter = false;
        x.SuppressInferBindingSourcesForParameters = true;
        x.InvalidModelStateResponseFactory = context => new InvalidModelStateResponseFactory(context).GetResponse();
      });

      services.Configure<Options>(Configuration);

    }

    public void Configure(IApplicationBuilder application, IWebHostEnvironment environment) {

      application.UseHsts();
      application.UseHttpsRedirection();

      application.UseRouting();
      application.UseCors("Cors");

      application.UseAuthentication();
      application.UseAuthorization();

      application.UseEndpoints(x => {
        x.MapControllers();
      });

    }

The **Auth Startup's** `ConfigureServices` and `Configure` methods are:

    public void ConfigureServices(IServiceCollection services) {

      services
        .AddControllersWithViews()
        .SetCompatibilityVersion(CompatibilityVersion.Latest)
        .AddJsonOptions()
        .AddFluentValidation();

      services.AddRouting(x => { x.AppendTrailingSlash = false; x.LowercaseUrls = true; });

      services.AddCors(x => {
        x.AddPolicy("Cors", y => 
          y.WithOrigins("https://localhost:5001", "https://localhost:5002").AllowAnyMethod().AllowAnyHeader().AllowCredentials());
        });

      services.AddDbContext<Context>(x => {

        x.UseSqlServer(Configuration.Get<Options>().Database.Connection, y => {
          y.MigrationsHistoryTable("__Migrations");
          y.UseNetTopologySuite();  
        }).EnableSensitiveDataLogging(Environment.IsDevelopment())
          .UseQueryTrackingBehavior(QueryTrackingBehavior.TrackAll);

        x.UseOpenIddict<Data.Entities.Application, Authorization, Scope, Token, Int32>();

      });

      services
        .AddIdentityCoreWithAuthentication<User>(x => x.AddDefaultOptions())
        .AddEntityFrameworkStores<Context>()
        .AddDefaultTokenProviders()
        .AddUserConfirmation<UserConfirmation<User>>();

      services.ConfigureApplicationCookie(x => {
        x.AccessDeniedPath = "/denied";
        x.Cookie.HttpOnly = false;
        x.Cookie.Name = "auth";
        x.ExpireTimeSpan = TimeSpan.FromMinutes(40);
        x.LoginPath = "/login";
        x.LogoutPath = "/logout";
        x.SlidingExpiration = true;
      });

      services.AddOpenIddict()
      
        .AddCore(x => {

          x.UseEntityFrameworkCore()
           .UseDbContext<Context>()
           .ReplaceDefaultEntities<Data.Entities.Application, Authorization, Scope, Token, Int32>();
        
        })

        .AddServer(x => {

          x.SetAuthorizationEndpointUris("/connect/authorize")
           .SetLogoutEndpointUris("/connect/logout")
           .SetTokenEndpointUris("/connect/token")
           .SetIntrospectionEndpointUris("/connect/introspect")
           .SetUserinfoEndpointUris("/connect/userinfo");

          x.RegisterClaims(OpenIddictConstants.Claims.Email, OpenIddictConstants.Claims.Name, OpenIddictConstants.Claims.Role);

          x.RegisterScopes(OpenIddictConstants.Scopes.Profile, OpenIddictConstants.Scopes.Email, OpenIddictConstants.Scopes.Roles, OpenIddictConstants.Scopes.OfflineAccess);
          
          x.AllowAuthorizationCodeFlow()
           .AllowRefreshTokenFlow();

      x.AddDevelopmentEncryptionCertificate().AddDevelopmentSigningCertificate();

          x.UseAspNetCore()
           .EnableAuthorizationEndpointPassthrough()
           .EnableLogoutEndpointPassthrough()
           .EnableTokenEndpointPassthrough()
           .EnableUserinfoEndpointPassthrough()
           .EnableStatusCodePagesIntegration();

        })

        .AddValidation(x => {
          x.UseLocalServer();
          x.UseAspNetCore();
        });

      services.AddHsts();
      services.AddHttpsRedirection();

      services.AddMediatR(typeof(Startup));

      services.Configure<Options>(Configuration);

      services.AddScoped<IUserClaimsPrincipalFactory<User>, UserClaimsPrincipalFactory>();

    } 

    public void Configure(IApplicationBuilder application, IWebHostEnvironment environment) {

      application.UseHsts();
      application.UseHttpsRedirection();

      application.UseStaticFiles();

      application.UseRouting();
      application.UseCors("Cors");

      application.UseAuthentication();
      application.UseAuthorization();

      application.UseEndpoints(x => {
        x.MapDefaultControllerRoute();
      });

    }

I've been trying a lot but always the same result.

### Further technical details
.NET SDK (reflecting any global.json):
 Version:   5.0.203
 Commit:    383637d63f

Runtime Environment:
 OS Name:     Mac OS X
 OS Version:  11.0
 OS Platform: Darwin
 RID:         osx.11.0-x64
 Base Path:   /usr/local/share/dotnet/sdk/5.0.203/

Host (useful for support):
  Version: 5.0.6
  Commit:  478b2f8c0e

.NET SDKs installed:
  5.0.203 [/usr/local/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 5.0.6 [/usr/local/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 5.0.6 [/usr/local/share/dotnet/shared/Microsoft.NETCore.App]

**Comment by kinosang** [comment]
Try with an API debug tools such as Postman and see how the OPTIONS pre-flight request handled, pre-flight response should have CORS headers and a 2xx status code. Also, the api it self should return CORS headers.

**Comment by kevinchalet** [comment]
Hi,

![image](https://user-images.githubusercontent.com/6998306/99829612-cf4f5300-2b5c-11eb-843b-d6b365965c0a.png)

As indicated, the ability to post questions on GitHub is reserved to sponsors as a way to thank them. Please consider [sponsoring  the project](https://github.com/sponsors/kevinchalet) or post your question on StackOverflow, where it will be visible by the community.

Thanks.

**Comment by mdmoura** [comment]
@kinosang I did used Insomnia API Client to try to find the error ... I was finally able to solve this:

It seems my code was fine ... I had a expired certificate on localhost. So run the following to solve it:

    dotnet dev-certs https --clean 
    dotnet dev-certs https --trust

**Comment by mdmoura** [comment]
> As indicated, the ability to post questions on GitHub is reserved to sponsors as a way to thank them. 

Yes, I am considering it as I am migrating from Identity Server to OpenIdDict. I am still trying it out ...

Would be useful to submit a demo using Authorization Code Flow with an Angular SPA acting as the client?
I might have some free time a few weeks after the launch of ASP.NET Core 6 ...

> Please consider [sponsoring the project](https://github.com/sponsors/kevinchalet) or post your question on StackOverflow, where it will be visible by the community.

I did posted a question on StacOverflow a few days before I created this issue.
When I created it I was convinced this could be a bug as the answers I got in StackOverflow found no error in the code.
